sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -305,12 +305,17 @@ object HiveTypeCoercion {
 
   /**
    * Convert all expressions in in() list to the left operator type
+   * except when the left operator type is NullType. In case when left hand
+   * operator type is NullType create a Literal(Null).
    */
   object InConversion extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveExpressions {
       // Skip nodes who's children have not been resolved yet.
       case e if !e.childrenResolved => e
 
+      case i @ In(a, b) if (a.dataType == NullType) =>
+        Literal.create(null, BooleanType)
```

Comment:
Ah sorry for my mistake. I thought you are casting the `a` to boolean type, but actually you just turn the result to boolean null.

Can you reference to a hive doc that says something like "if the value is null, the In operation will always return null"? I think we should follow hive semantic here.